### PR TITLE
Add DirectoryConfigProvider to the service provider list

### PR DESCRIPTION
**Pull request by scholzj**
Following the discussion in confluentinc/kafka-images#102, it seems that the `DirectoryConfigProvider` should be also listed in the list of the service providers.

CC: @C0urante

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by C0urante — APPROVED** [review]
Thanks @scholzj 👍

**Review by tombentley — APPROVED** [review]

**Comment by tombentley** [comment]
I think we should add a line to the class javadoc for `ConfigProvider`, because the requirement for `ServiceLoader` discoverability doesn't seem to be documented (at least, that I know of) and is easily missed.
Also, we should add a test for discovering via `ServiceLoader`. As well as regression, people are likely to look at these classes and tests when writing their own config providers, so having a test for it is more likely to ensure they don't make the same mistake.

**Comment by scholzj** [comment]
@tombentley I added the JavaDoc as well as tests for both providers.

File: clients/src/main/java/org/apache/kafka/common/config/provider/ConfigProvider.java [hunk 1]
```diff
@@ -26,6 +26,8 @@
 /**
  * A provider of configuration data, which may optionally support subscriptions to configuration changes.
  * Implementations are required to safely support concurrent calls to any of the methods in this interface.
+ * Configuration provider implementations are in Kafka Connect loaded as Service Providers using the {@code ServiceLoader} class.
+ * To support this, implementations of this interface should also contain its service provider configuration file.
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/common/config/provider/ConfigProvider.java]:
```suggestion
 * Kafka Connect discovers configuration providers using Java's Service Provider mechanism (see {@code java.util.ServiceLoader}).
 * To support this, implementations of this interface should also contain a service provider configuration file in {@code  META-INF/service/org.apache.kafka.common.config.provider.ConfigProvider}.
```

File: clients/src/test/java/org/apache/kafka/common/config/provider/DirectoryConfigProviderTest.java [hunk 2]
```diff
@@ -145,5 +146,21 @@ public void testNullPathWithKey() {
...
+        boolean discovered = false;
+
+        for (ConfigProvider service : serviceLoader)    {
+            System.out.println(service.getClass());
+            if (service instanceof DirectoryConfigProvider) {
+                discovered = true;
+            }
+        }
+
+        assertTrue(discovered);
```
Comment (tombentley) [inline clients/src/test/java/org/apache/kafka/common/config/provider/DirectoryConfigProviderTest.java]:
```suggestion
        assertTrue(StreamSupport.stream(serviceLoader.spliterator(), false).anyMatch(sl -> sl instanceof DirectorConfigProvider));
```

File: clients/src/test/java/org/apache/kafka/common/config/provider/FileConfigProviderTest.java [hunk 3]
```diff
@@ -87,6 +88,22 @@ public void testNullPathWithKey() throws Exception {
...
+        boolean discovered = false;
+
+        for (ConfigProvider service : serviceLoader)    {
+            System.out.println(service.getClass());
+            if (service instanceof FileConfigProvider) {
+                discovered = true;
+            }
+        }
+
+        assertTrue(discovered);
```
Comment (tombentley) [inline clients/src/test/java/org/apache/kafka/common/config/provider/FileConfigProviderTest.java]: Same refactoring

File: clients/src/test/java/org/apache/kafka/common/config/provider/FileConfigProviderTest.java [hunk 4]
```diff
@@ -87,6 +89,13 @@ public void testNullPathWithKey() throws Exception {
         assertNull(configData.ttl());
     }
 
+    @Test
+    public void testServiceLoaderDiscovery() {
+        ServiceLoader<ConfigProvider> serviceLoader = ServiceLoader.load(ConfigProvider.class);
+        assertTrue(StreamSupport.stream(serviceLoader.spliterator(), false).anyMatch(sl -> sl instanceof FileConfigProvider));
+
```
Comment (tombentley) [inline clients/src/test/java/org/apache/kafka/common/config/provider/FileConfigProviderTest.java]: Blank line

File: clients/src/test/java/org/apache/kafka/common/config/provider/DirectoryConfigProviderTest.java [hunk 5]
```diff
@@ -145,5 +147,11 @@ public void testNullPathWithKey() {
         assertTrue(configData.data().isEmpty());
         assertNull(configData.ttl());
     }
+
+    @Test
+    public void testServiceLoaderDiscovery() {
+        ServiceLoader<ConfigProvider> serviceLoader = ServiceLoader.load(ConfigProvider.class);
+        assertTrue(StreamSupport.stream(serviceLoader.spliterator(), false).anyMatch(sl -> sl instanceof DirectoryConfigProvider));
```
Comment (tombentley) [inline clients/src/test/java/org/apache/kafka/common/config/provider/DirectoryConfigProviderTest.java]: `sl` -> `configProvider`

File: clients/src/test/java/org/apache/kafka/common/config/provider/FileConfigProviderTest.java [hunk 6]
```diff
@@ -87,6 +89,13 @@ public void testNullPathWithKey() throws Exception {
         assertNull(configData.ttl());
     }
 
+    @Test
+    public void testServiceLoaderDiscovery() {
+        ServiceLoader<ConfigProvider> serviceLoader = ServiceLoader.load(ConfigProvider.class);
+        assertTrue(StreamSupport.stream(serviceLoader.spliterator(), false).anyMatch(sl -> sl instanceof FileConfigProvider));
```
Comment (tombentley) [inline clients/src/test/java/org/apache/kafka/common/config/provider/FileConfigProviderTest.java]: `sl` -> `configProvider`

**Comment by tombentley** [comment]
Failing tests are unrelated.

**Pull request merged**